File: PA34.java
//LeetCode:15 3sum 
//Time Complexity : O(n^2) // O(nlogn+ n^2) where n is number of elements in nums
//Space Complexity : O(1) // List is auxillary space
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        
        if(nums== null | nums.length<3){
            return new ArrayList<List<Integer>>();
        }
        List<List<Integer>> res= new ArrayList<List<Integer>>();
        Arrays.sort(nums);
        List<Integer> ls;
        int sum=0;
        for(int i=0; i<nums.length-2;i++){
            
            int low= i+1;
            int high =nums.length-1;
            if(i>0 && nums[i]==nums[i-1]) continue;
      
            while(low<high){
               
                sum=nums[i]+nums[low]+nums[high];
                if(sum==0){
                    ls= new ArrayList<Integer>();
                    ls.add(nums[i]);
                    ls.add(nums[low]);
                    ls.add(nums[high]);
                    res.add(ls);
                    high--;
                    low++;
                    while(low<high && nums[low]==nums[low-1]) low++;
                    while(low<high && nums[high]==nums[high+1]) high--;
                }
                else{
                    if(sum<0){
                    low++;
                }
                else{
                    high--;
                }
                }
                
            }
     
        }
        return res;
    }
}

File: PA_34.py
#LC 15 - 3Sum
#Time Complexity - O(n^2)
#Space Complexity - O(1)
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        if len(nums) < 3:
            return []
        i = 0
        nums = sorted(nums)
        ans = []
        #print(nums)
        for i in range(len(nums)-2):
            if i != 0 and nums[i-1] == nums[i]:
                continue
            if nums[i] > 0:
                break
            low = i + 1
            high = len(nums) - 1
            while low < high:
                tsum = nums[i] + nums[low] + nums[high]
                if tsum == 0:
                    #print(i,low,high)
                    ans.append([nums[i],nums[low],nums[high]])
                    while low < high and nums[low] == nums[low-1]:
                        low = low + 1
                    while low < high and nums[high] == nums[high-1]:
                        high = high - 1
                    high = high - 1
                elif tsum < 0:
                    low = low + 1
                else:
                    high = high - 1
        
        return ans